insonmnia/resource/pool.go

```diff
@@ -1,116 +1,187 @@
+	defer m.mu.Unlock()
+	return m.pool.getFree()
+}
+
+func (m *Scheduler) CGroup(askPlanID string) (cgroups.CGroup, error) {
+	panic("implement me")
 }
 
-func (p *Pool) GetUsage() Resources {
-	p.mu.Lock()
-	defer p.mu.Unlock()
+func (m *Scheduler) PollConsume(askPlan *sonm.AskPlan) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
 
-	return p.usage
+	return m.pool.pollConsume(askPlan.Resources)
 }
 
 // Consume tries to consume the specified resource usage from the pool.
 //
 // Does nothing on error.
-func (p *Pool) Consume(usage *Resources) error {
-	p.mu.Lock()
-	defer p.mu.Unlock()
+func (m *Scheduler) Consume(askPlan *sonm.AskPlan) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
 
-	return p.consume(usage)
+	return m.pool.consume(askPlan.ID, askPlan.Resources)
 }
 
-func (p *Pool) consume(usage *Resources) error {
-	if err := p.pollConsume(usage); err != nil {
-		return err
+func (m *Scheduler) ConsumeTask(askPlanID string, taskID string, resources *sonm.AskPlanResources) error {
+	copy := &sonm.AskPlanResources{
+		GPU:     deepcopy.Copy(resources.GPU).(*sonm.AskPlanGPU),
+		Storage: deepcopy.Copy(resources.Storage).(*sonm.AskPlanStorage),
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	m.taskToAskPlan[taskID] = askPlanID
+	pool, ok := m.askPlanPools[askPlanID]
+	if !ok {
+		return fmt.Errorf("could not consume resources for task - ask Plan with id %s not found", askPlanID)
 	}
 
-	p.usage.NumCPUs += usage.NumCPUs
-	p.usage.Memory += usage.Memory
-	if usage.NumGPUs == -1 {
-		p.usage.NumGPUs = len(p.OS.GPU)
-	} else {
-		p.usage.NumGPUs += usage.NumGPUs
+	return pool.consume(taskID, copy)
+}
+
+func (m *Scheduler) Release(askPlanID string) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	return m.pool.release(askPlanID)
+}
+
+func (m *Scheduler) ReleaseTask(taskID string) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	askPlanID, ok := m.taskToAskPlan[taskID]
+	if !ok {
+		return fmt.Errorf("could not find corresponding ask plan id for task %s", taskID)
+	}
+	pool, ok := m.askPlanPools[askPlanID]
+	if !ok {
+		return fmt.Errorf("could not release resources for task - ask Plan with id %s not found", askPlanID)
 	}
 
+	err := pool.release(taskID)
+	if err != nil {
+		return err
+	}
+	delete(m.taskToAskPlan, taskID)
 	return nil
 }
 
-func (p *Pool) PollConsume(usage *Resources) error {
-	p.mu.Lock()
-	defer p.mu.Unlock()
+type pool struct {
+	all  *sonm.AskPlanResources
+	used map[string]*sonm.AskPlanResources
+}
 
-	return p.pollConsume(usage)
+func newPool(resources *sonm.AskPlanResources) *pool {
+	return &pool{
+		all:  resources,
+		used: map[string]*sonm.AskPlanResources{},
+	}
 }
 
-func (p *Pool) pollConsume(usage *Resources) error {
-	if usage.NumGPUs == -1 {
-		usage.NumGPUs = len(p.OS.GPU)
+func (p *pool) getFree() (*sonm.AskPlanResources, error) {
+	res := deepcopy.Copy(p.all).(*sonm.AskPlanResources)
+	usage, err := p.getUsage()
+	if err != nil {
+		return nil, err
 	}
+	err = res.Sub(usage)
+	if err != nil {
+		pool, _ := yaml.Marshal(res)
```

Comment:
Why to ignore errors?

Reply:
because it is marshalling for error printing (and this should never fail)

Reply:
But it can fail.

Reply:
I really don't get it, what is your suggestion? This is used only to format error and in worst case we will have badly formatted error and that's all. btw it can fail only if some non-marshalable structure is passed to yaml.Marshal, which is not our case - overvise that would indicate logic error. I can put panic there of course if you want